KIALI-3193 Add support for ServiceMeshPolicy,ServiceMeshRbacConfig

lucasponce opened the pull request:
** Describe the change **

Maistra distribution will clone MeshPolicy and ClusterRbacConfig to ServiceMeshPolicy and ServiceMeshRbacConfig objects with identical meaning but these objects are namespaced.

jotak commented on kubernetes/istio_details_service.go:
```diff
@@ -155,6 +158,7 @@ func (in *IstioClient) getRbacResources() map[string]bool {
 	}
 
 	rbacResources := map[string]bool{}
+	// TODO Update with Maistra API
```
Is it to do within this PR ?

lucasponce replied:
Yeah. I'll update that part.

lucasponce replied:
I've reviewed and updated the code.
Basically the comment was not correct as the right logic to check if it needs to get a ClusterRbacConfig or a ServiceMeshRbacConfig is handled in the upper methods "GetIstioConfigList" and "GetAuthorizationDetails".

lucasponce commented:
I've added the backport label to not forget, but I don't think this PR can be cherry-picked, so I'm going to prepare proper PRs for v1.0 branch as well.

lucasponce commented:
@hhovsepy I missed to update the operator with the new permissions.
Sorry, the PR is not yet ready. Probably you will confirm that it fails when deployed.

lucasponce commented:
@mwringe @jmazzitelli note that this PR and the version fo v1.0 branch in #1327 will update the permissions of Kiali.
I guess those needs to be propagated also properly with Maistra integration but I don't know where.
Can you take care of that part when this PR and #1327 are ready ?

jmazzitelli commented:
> those needs to be propagated also properly

See: https://issues.jboss.org/browse/KIALI-3203

jmazzitelli commented on operator/roles/kiali-deploy/templates/openshift/role.yaml:
```diff
@@ -130,6 +130,26 @@ rules:
   - list
   - patch
   - watch
+- apiGroups: ["authentication.maistra.io"]
```
these new permissions need to go in the operator's own role - otherwise the operator will not be able to install kiali.

So add to this file in this spot to remain consistent: https://github.com/kiali/kiali/blob/master/operator/deploy/role.yaml#L260

jotak replied:
I wonder ( @jmazzitelli , @lucasponce ) : are these changes with explicit mention to maistra fine even in a non-maistra setup, but still on openshift ?
I guess it's still going to work fine when "rbac.maistra.io" isn't anything on the cluster, but prefer to ask

jmazzitelli replied:
yeah, these api groups don't have to exist. they will be silently ignored. At least, that's what I have seen.

hhovsepy commented:
`FAILED! => {"attempts": 6, "changed": false, "error": 403, "msg": "Failed to create object: {\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"clusterroles.rbac.authorization.k8s.io \\\"kiali\\\" is forbidden: user \\\"system:serviceaccount:kiali-operator:kiali-operator\\\" (groups=[\\\"system:serviceaccounts\\\" \\\"system:serviceaccounts:kiali-operator\\\" \\\"system:authenticated\\\"]) is attempting to grant RBAC permissions not currently held:\\n{APIGroups:[\\\"authentication.maistra.io\\\"], Resources:[\\\"servicemeshpolicies\\\"], Verbs:[\\\"create\\\" \\\"delete\\\" \\\"get\\\" \\\"list\\\" \\\"patch\\\" \\\"watch\\\"]}\\n{APIGroups:[\\\"rbac.maistra.io\\\"], Resources:[\\\"servicemeshrbacconfigs\\\"], Verbs:[\\\"create\\\" \\\"delete\\\" \\\"get\\\" \\\"list\\\" \\\"patch\\\" \\\"watch\\\"]}\",\"reason\":\"Forbidden\",\"details\":{\"name\":\"kiali\",\"group\":\"rbac.authorization.k8s.io\",\"kind\":\"clusterroles\"},\"code\":403}\n", "reason": "Forbidden", "status": 403}`

jmazzitelli commented:
Yes, @hhovsepy - that failure is expected because this is still missing from this PR: https://github.com/kiali/kiali/pull/1326#pullrequestreview-273266994

jmazzitelli commented:
@lucasponce  I create a PR to your PR with the necessary change to the operator's role.yaml here: https://github.com/lucasponce/kiali/pull/2/files#diff-e591ae4b056dba37ffd55d13feab9ee5

lucasponce commented:
@jmazzitelli let me rebase my branch, I'm reviewing your PR to https://github.com/lucasponce/kiali/pull/2/files#diff-e591ae4b056dba37ffd55d13feab9ee5 and I see a lot of noise with the commits.

I will have it ready shortly, so when you are online, we can sync.

lucasponce commented:
@hhovsepy @jmazzitelli I've rebased and added the operator permissions on the PR.
Let me know if it's looking good now.
I'm going to prepare the v1.0 PR.

hhovsepy commented:
Still getting the same error: https://github.com/kiali/kiali/pull/1326#issuecomment-520029252
Deployed on OCP 4.1 and Maistra latest-qe build.

lucasponce commented:
@hhovsepy can you test on 3.11 as well ?
Perhaps here for 4.1 we need some additional piece related to OLM.
@jmazzitelli any idea about error Hayk is reporting ?

gbaufake commented:
@jmazzitelli 
Is this related to removal of clusterroles of kiali-operator?

hhovsepy commented:
@lucasponce on 3.11 as well, the same error in kiali-operator logs.

jmazzitelli commented:
Yes, you need to give the operator those role permissions. @hhovsepy I don't know how QE is creating/deploying the operator, but if you guys have your own deployment scripts, you need to add those new permissions. If you are using the deploy-kiali-operator.sh script, you need to make sure you are using the one in this PR.

You can see the operator doesn't have these new permissions that are required - from the reported error:
```
is attempting to grant RBAC permissions not currently held:\\n{APIGroups:[\\\"authentication.maistra.io\\\"], Resources:[\\\"servicemeshpolicies\\\"], Verbs:[\\\"create\\\" \\\"delete\\\" \\\"get\\\" \\\"list\\\" \\\"patch\\\" \\\"watch\\\"]}\\n{APIGroups:[\\\"rbac.maistra.io\\\"], Resources:[\\\"servicemeshrbacconfigs\\\"], Verbs:[\\\"create\\\" \\\"delete\\\" \\\"get\\\" \\\"list\\\" \\\"patch\\\" 
```

gbaufake commented:
@jmazzitelli 
We have been using the uber.yaml spit from deploy-operator.sh script
[kiali-operator.yaml.txt](https://github.com/kiali/kiali/files/3492412/kiali-operator.yaml.txt)

 Any changes that need to be done here?

jmazzitelli commented:
> Any changes that need to be done here?

Yes.. any time we add changes to the operator's deploy yaml, you will need to update this. In this case, new role permissions were added. You will need to add those to your uber yaml. This explains the errors @hhovsepy is seeing.

lucasponce commented:
@hhovsepy I could test these pair of PRs on my 3.11 using Maistra latest-qe with this config:
https://gist.github.com/lucasponce/5654eca6ac6e10c96014e3075f33e2f4
![image](https://user-images.githubusercontent.com/1662329/62866748-eaa0a100-bd11-11e9-9c10-11b707fcdcc2.png)
![image](https://user-images.githubusercontent.com/1662329/62866770-f8562680-bd11-11e9-85ac-071e598bd31d.png)

lucasponce commented:
@hhovsepy ops
found a problem that affects to the validations visualization.
I will send a commit shortly.
